Add Boiler Efficiency output for Boiler objects

matthew-larson opened the pull request:
Pull request overview
---------------------
 - Fixes #8064 
 - This pull request adds output variables for Boiler (Steam) Efficiency for the Boiler:HotWater and Boiler:Steam objects.  This provides an efficiency output variable for the Boiler objects similar to other plant objects like the Chiller objects providing a Chiller COP output variable.  For simplification purposes, I changed the variable Effic for boiler efficiency at design conditions to NomEffic, similar to NomCap, and created a new variable BoilerEff to be the outputted variable for each timestep.  Unit tests were created to ensure the BoilerEff variable result is correct.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts

matthew-larson committed:
Add BoilerEff varaible as an output variable and change original BoilerEff to BoilerNomEff for clarification.  Add output variable to documentation as well.

matthew-larson committed:
Add BoilerEff to default constructor.

matthew-larson committed:
Add boiler efficiency for steam boilers.

matthew-larson committed:
Merge branch 'develop' of https://github.com/nrel/energyplus into add-boilereff-output

matthew-larson committed:
Add unit test for boiler efficiency and it's output.

matthew-larson committed:
Modify LoopDemandCalcScheme if statment to match Boilers.cc and prevent unnecessary crash in unit test.

matthew-larson committed:
Add unit test to check boiler steam efficiency output variable.

mitchute commented on doc/input-output-reference/src/overview/group-plant-equipment.tex:
```diff
@@ -4159,6 +4161,10 @@ \subsubsection{Outputs}\label{outputs-10-005}
 This output is the operating part-load ratio of the boiler. The part-load ratio is calculated as the boiler load divided by the boiler nominal capacity. The part-load ratio is limited by the minimum and maximum part-load ratio inputs specified by the user.
 
+\paragraph{Boiler Efficiency {[]}}\label{boiler-efficiency}
+
```
I like the note, but just for the sake of uniformity and completeness, we should probably also have an explicit definition of efficiency. Maybe add something like "this output is boiler heating rate divided by boiler gas rate" or similar between the label and the note.

mitchute commented on doc/input-output-reference/src/overview/group-plant-equipment.tex:
```diff
@@ -4401,6 +4409,10 @@ \subsubsection{Outputs}\label{outputs-11-004}
 These outputs are the heating output (load) of the boiler. Energy is metered on Boilers:EnergyTransfer, EnergyTransfer:Plant, and EnergyTransfer:Facility.
 
+\paragraph{Boiler Steam Efficiency {[]}}\label{boiler-steam-efficiency}
+
```
Same change here.

mitchute commented on src/EnergyPlus/BoilerSteam.cc:
```diff
@@ -214,7 +214,7 @@ namespace BoilerSteam {
                 ShowContinueError("Field: Maximum Operation Pressure units are Pa. Verify units.");
             }
-            thisBoiler.Effic = DataIPShortCuts::rNumericArgs(2);
+            thisBoiler.NomEffic = DataIPShortCuts::rNumericArgs(2);
```
More explicit. This is good.

mitchute commented on tst/EnergyPlus/unit/BoilerHotWater.unit.cc:
```diff
@@ -185,3 +191,95 @@ TEST_F(EnergyPlusFixture, Boiler_HotWater_BlankDesignWaterFlowRate)
     // Additional tests for fuel type input
     EXPECT_EQ(state.dataBoilers.Boiler(1).BoilerFuelTypeForOutputVariable, "Gas");
 }
+
```
Unit tests look good.

mitchute reviewed (changes requested):
Minor doc changes requested in the IORef but this looks good other than that. Running tests locally now.

mitchute commented:
Unit tests look good, and I ran a few files and the results look good. We also need to add the new output variables to a least one file for the Boiler:HotWater and Boiler:Steam objects.

matthew-larson committed:
Merge branch 'develop' of https://github.com/nrel/energyplus into add-boilereff-output

matthew-larson committed:
add definition of boiler efficiency in documentation.

matthew-larson committed:
Add boiler efficiency output variables to test files.

matthew-larson commented:
@mitchute I added descriptions for the output variables as requested and also added the output variables to a few files with Boiler:HotWater and Boiler:Steam objects that were outputting similar variables.

mitchute commented:
Thanks, @matthew-larson. This should go in once CI reports back.

matthew-larson committed:
Fix build failures for unit tests.

matthew-larson commented:
Just pushed up changes to fix the build errors.

mitchute commented:
AUD and RDD diffs are expected here. Merging.
